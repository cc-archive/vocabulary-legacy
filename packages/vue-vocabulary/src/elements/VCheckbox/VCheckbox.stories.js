import { figmaConfig } from '@/utils/helpers'
import VCheckbox from '@/elements/VCheckbox/VCheckbox'

const checkboxDescription = 'The Checkbox component is an essential forms component.'

export default {
  title: 'Elements/VCheckbox',
  parameters: { design: figmaConfig('603%3A3'),
    docs: { description: { component: checkboxDescription } } },
  component: VCheckbox
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { VCheckbox },
  template: `
    <div class="checkbox-container">
    <h3>Checkbox selected values</h3>
    <div>
      <p>Default checkbox: <b>{{defaultCheckbox}}</b></p>
      <p>Checked checkbox: <b>{{checkedCheckbox}}</b></p>
      <p>Disabled checkbox: <b>{{disabledCheckbox}}</b></p>
      <p>Disabled Checked checkbox: <b>{{disabledCheckedCheckbox}}</b></p>
      <p>Long checkbox: <b>{{longCheckbox}}</b></p>
    </div>
    <v-checkbox v-model="defaultCheckbox">Default Option</v-checkbox>
    <v-checkbox v-model="checkedCheckbox">Checked Option</v-checkbox>
    <v-checkbox v-model="disabledCheckbox" :disabled="true">Disabled Option</v-checkbox>
    <v-checkbox v-model="disabledCheckedCheckbox" :disabled="true">Disabled Checked Option</v-checkbox>
    <v-checkbox v-model="longCheckbox">
      This is a long label for agreement with conditions of a license. Lorem ipsum dolor
      sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore
      et dolore magna aliqua.
      Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip
      ex ea commodo consequat.</v-checkbox>
    </div>
    `,
  data: () => {
    return {
      defaultCheckbox_: undefined,
      checkedCheckbox_: true,
      disabledCheckbox_: undefined,
      disabledCheckedCheckbox_: true,
      longCheckbox_: undefined
    }
  },
  computed: {
    longCheckbox: {
      get () {
        return this.longCheckbox_
      },
      set (newVal) {
        this.longCheckbox_ = newVal
      }
    },
    disabledCheckedCheckbox: {
      get () {
        return this.disabledCheckedCheckbox_
      },
      set (newVal) {
        this.disabledCheckedCheckbox_ = newVal
      }
    },
    disabledCheckbox: {
      get () {
        return this.disabledCheckbox_
      },
      set (newVal) {
        this.disabledCheckbox_ = newVal
      }
    },
    checkedCheckbox: {
      get () {
        return this.checkedCheckbox_
      },
      set (newVal) {
        this.checkedCheckbox_ = newVal
      }
    },
    defaultCheckbox: {
      get () {
        return this.defaultCheckbox_
      },
      set (newVal) {
        this.defaultCheckbox_ = newVal
      }
    }
  }
})
export const Default = Template.bind({})
