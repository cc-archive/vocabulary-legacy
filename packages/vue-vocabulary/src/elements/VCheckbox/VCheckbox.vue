<template>
    <label class="v-checkbox checkbox">
      <input type="checkbox" :checked="value" :disabled="disabled" v-model="computedValue">
      <span class="checkbox-label">
        <slot default />
      </span>
    </label>
</template>

<script>
  export default {
    name: 'VCheckbox',
    props: {
      /**
       * Checkbox is checked if this is true. This value is set by `v-model`.
       */
      value: {
        type: [ Boolean, undefined ],
        default: undefined
      },
      /**
       * Optional property, set to `true` to disable the checkbox
       */
      disabled: {
        type: Boolean,
        default: false
      }
    },
    data () {
      return {
        localValue: this.value
      }
    },
    computed: {
      /**
       * Emitting input event
       */
      computedValue: {
        get () {
          return this.localValue
        },
        set (value) {
          this.localValue = value
          this.$emit('input', value)
        }
      }
    },
    watch: {
      /**
       * When v-model change, set internal value.
       */
      value (value) {
        this.localValue = value
      }
    },
    methods: {
      handleClick () {
        console.log('Doing something')
      },
      focus () {
        // MacOS FireFox and Safari do not focus when clicked
        this.$refs.input.focus()
      }
    }
  }
</script>

<style lang="scss" scoped>
@import "~@creativecommons/vocabulary/scss/color.scss";
.checkbox, .radio {
  display: flex;
  align-items: baseline;
  color: #{$color-dark-slate-gray};
}

// This is overriding Bulma style for horizontal radio elements?
.radio+.radio {
  margin-left: unset;
}
@supports(-webkit-appearance: none) or (-moz-appearance: none) {
  input[type='checkbox'],
  input[type='radio'] {
    margin: 0 0.5rem 0 0;
    flex: 0 0 auto;
    width: 1rem;
    height: 1rem;
    -webkit-appearance: none;
    -moz-appearance: none;
    --checkmark-color: #fff;

    outline: none;
    display: inline-block;
    vertical-align: top;
    position: relative;
    cursor: pointer;
    border: 1px solid var(--border-color, #{$color-light-gray});
    background: var(--background);
    transition: background .3s, border-color .3s, box-shadow .2s;

    &:after {
      content: '';
      display: block;
      left: 0;
      top: 0;
      position: absolute;
      transition: transform .3s ease, opacity .2s;
      opacity: var(--o, 0);
    }

    &:checked {
      --background: #{$color-brighter-blue};
      --border-color: transparent;
      --o: 1;

      &:not(:disabled):hover {
        filter: brightness(120%) saturate(120%);
      }
    }

    &:disabled {
      --background: #{$color-light-gray};
      --border-color: #{$color-gray};
      cursor: not-allowed;

      &:checked {
        --checkmark-color: #{$color-gray};
      }
    }

    &:focus {
      box-shadow: 0 0 0 2px rgba(39, 94, 254, .3);
    }

    &:hover:not(:checked):not(:disabled) {
      --border-color: #{$color-gray};
    }
  }
  input[type='checkbox'] {
    border-radius: 0.2rem;

    &:after {
      width: 5px;
      height: 9px;
      border: 2px solid var(--checkmark-color);
      border-top: 0;
      border-left: 0;
      left: 4px;
      top: 2px;
      transform: rotate(20deg);
    }

    &:checked::after {
      transform: rotate(35deg);
    }
  }
  input[type='radio'] {
    border-radius: 0.5rem;

    &:checked::after {
      background: var(--checkmark-color);
      width: 0.5rem;
      height: 0.5rem;
      border-radius: 0.25rem;
      left: 3px;
      top: 3px;
    }

    &:disabled {
      --checkmark-color: #{$color-gray};

      & + .radio-label {
        color: #{$color-gray};
      }
    }
  }
}
</style>
