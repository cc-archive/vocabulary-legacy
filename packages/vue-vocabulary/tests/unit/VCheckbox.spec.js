import VCheckbox from '@/elements/VCheckbox/VCheckbox'
import { mount } from '@vue/test-utils'
describe('VCheckbox', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mount(VCheckbox, {
      slots: {
        default: 'Default Checkbox'
      }
    })
  })
  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('emits "true" when checked', () => {
    const checkbox = wrapper.find('input[type="checkbox"]')
    checkbox.setChecked()
    expect(wrapper.emitted().input[0]).toEqual([true])
  })
  it('emits "false" when unchecked', async () => {
    await wrapper.setProps({
      value: true
    })
    const checkbox = wrapper.find('input[type="checkbox"]')
    expect(checkbox.element.checked).toBeTruthy()
    await checkbox.element.click()
    expect(wrapper.emitted().input[0]).toEqual([false])
  })
  it('cannot be clicked when disabled', async () => {
    await wrapper.setProps({
      value: true,
      disabled: true
    })
    const checkbox = wrapper.find('input[type="checkbox"]')
    expect(checkbox.element.checked).toBeTruthy()
    await checkbox.element.click()
    expect(wrapper.emitted().input).toBeFalsy()
  })
})
