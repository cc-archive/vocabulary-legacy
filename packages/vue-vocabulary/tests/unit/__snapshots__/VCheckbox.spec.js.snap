// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VCheckbox matches snapshot 1`] = `<label class="v-checkbox checkbox"><input type="checkbox"> <span class="checkbox-label">Default Checkbox</span></label>`;
